Speed up taint mode by memoizing range computation for expressions

IagoAbal opened the pull request:
Closes #3596

test plan:
semgrep-core -lang py -profile \
    -config semgrep-rules/python/django/security/injection/open-redirect.yaml \
    semgrep/perf/bench/django/input/django
    #^ This is now as fast as the old fake-taint rule and range_of_any_opt
    # amounts for less than 20% of the total time (vs above 60% before).



PR checklist:
- [x] documentation is up to date
- [x] changelog is up to date

aryx commented on semgrep-core/src/core/ast/AST_generic.ml:
```diff
@@ -381,7 +381,14 @@ and id_info = {
+and expr = {
+  e : expr_kind;
+  e_id : int;
+  (* used in semgrep to quickly get the range of a matched statement *)
```
typo: of a matched expression

aryx commented on semgrep-core/src/core/ast/AST_generic.ml:
```diff
@@ -381,7 +381,14 @@ and id_info = {
+  mutable e_range :
+    (Parse_info.token_location * Parse_info.token_location) option;
+      [@hash.ignore]
+}
```
Probably need to also define the equal for it. I think we forgot for s_range but it's a bug, we should.

aryx commented:
CI failures, you probably need to make regenerate-tests. We have some debugging test that
do not abstract away the md5sum and so fail each time we change the expression type.
Not sure who wrote the debugging test, but I remember @brendongo or @mschwager wrote some code to not care about those md5sum in the expected json output.

IagoAbal committed:
core: Memoize range_of_any_opt for expressions

Closes #3596

test plan:
semgrep-core -lang py -profile \
    -config semgrep-rules/python/django/security/injection/open-redirect.yaml \
    semgrep/perf/bench/django/input/django
    #^ This is now as fast as the old fake-taint rule and range_of_any_opt
    # amounts for less than 20% of the total time (vs above 60% before).
